stonesoup/base.py

```diff
@@ -305,3 +307,22 @@ def __repr__(self):
         params = ("{}={!r}".format(name, getattr(self, name))
                   for name in type(self).properties)
         return "{}({})".format(type(self).__name__, ", ".join(params))
+
+    def __deepcopy__(self, memodict={}):
+        # Create a new class
+        new = object.__new__(type(self))
+        memodict[id(self)] = new   # add the new class to the memo
+        # Insert a deepcopy of all instance attributes
+        new.__dict__.update(deepcopy(self.__dict__, memodict))
+        # Manually update any weakrefs to point to copies, if they exist.
+        for name, prop in new.__dict__.items():
+            if isinstance(prop, weakref.ref):
+                original_target = prop()  # call the weakref to get a reference
+                try:
+                    # if we are copying the parent as well, the copy should be in memodict
+                    copy_of_target = memodict[id(original_target)]
+                    new.__setattr__(name, weakref.ref(copy_of_target))
+                except KeyError:
+                    # if we can't find the parent, then leave the original ref in place
+                    pass
```

Comment:
Minor suggestion.

```suggestion
                try:
                    # if we are copying the parent as well, the copy should be in memodict
                    copy_of_target = memodict[id(original_target)]
                except KeyError:
                    # if we can't find the parent, then leave the original ref in place
                    pass
                else:
                    new.__setattr__(name, weakref.ref(copy_of_target))
```

Reply:
Not sure that will work. I don't think `copy_of_target` will be set in the `KeyError` case.

Reply:
`else` block is only executed if there isn't an error. This is useful to avoid unintentionally catching another `KeyError` when calling `__setattr__`

Reply:
Sorry, was mistaking for `finally`. Haven't seen that usage. You learn something every day!

P.S. This is why everyone should both have and do code reviews :)